File: src/kernel/schemas/scheduled_task.py
"""Scheduled task schemas."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field

# ── Enums ──────────────────────────────────────────


class TriggerType(str, Enum):
    INTERVAL = "interval"
    CRON = "cron"
    DATE = "date"


class ScheduledTaskStatus(str, Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    DELETED = "deleted"


class RunStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    FAILED = "failed"
    SKIPPED = "skipped"
    TIMEOUT = "timeout"


# ── Trigger configs ────────────────────────────────


class IntervalTriggerConfig(BaseModel):
    """Fixed-interval trigger."""

    seconds: int = Field(..., ge=1, description="Interval in seconds")


class CronTriggerConfig(BaseModel):
    """Cron-expression trigger. All fields accept standard cron syntax."""

    year: Optional[str] = Field(None, description="Year pattern")
    month: Optional[str] = Field(None, description="Month pattern (1-12)")
    day: Optional[str] = Field(None, description="Day of month pattern (1-31)")
    week: Optional[str] = Field(None, description="ISO week pattern (1-53)")
    day_of_week: Optional[str] = Field(None, description="Day of week pattern (mon,tue,...)")
    hour: Optional[str] = Field("0", description="Hour pattern (0-23)")
    minute: Optional[str] = Field("0", description="Minute pattern (0-59)")
    second: Optional[str] = Field("0", description="Second pattern (0-59)")


class DateTriggerConfig(BaseModel):
    """One-time trigger at a specific UTC timestamp."""

    run_date: datetime = Field(..., description="One-time execution datetime")


# ── Task models ────────────────────────────────────


class ScheduledTaskCreate(BaseModel):
    """Request body for creating a scheduled task."""

    name: str = Field(..., min_length=1, max_length=200)
    agent_id: str = Field(..., min_length=1)
    trigger_type: TriggerType
    trigger_config: dict = Field(
        ...,
        description="Trigger config (IntervalTriggerConfig | CronTriggerConfig | DateTriggerConfig)",
    )
    input_payload: dict = Field(default_factory=dict, description="Agent input parameters")
    description: Optional[str] = Field(None, max_length=2000)
    enabled: bool = Field(True)
    run_on_start: bool = Field(False)
    max_retries: int = Field(0, ge=0, le=10)
    timeout_seconds: int = Field(600, ge=10, le=3600)
    source_session_id: Optional[str] = Field(
        None, description="Conversation session where the task was created"
    )
    source_run_id: Optional[str] = Field(None, description="Agent run where the task was created")
    created_by: str = Field("user", description="Creator source: user / agent / api")


class ScheduledTaskUpdate(BaseModel):
    """Request body for updating a scheduled task."""

    name: Optional[str] = Field(None, min_length=1, max_length=200)
    agent_id: Optional[str] = Field(None, min_length=1)
    trigger_type: Optional[TriggerType] = None
    trigger_config: Optional[dict] = None
    input_payload: Optional[dict] = None
    description: Optional[str] = Field(None, max_length=2000)
    enabled: Optional[bool] = None
    run_on_start: Optional[bool] = None
    max_retries: Optional[int] = Field(None, ge=0, le=10)
    timeout_seconds: Optional[int] = Field(None, ge=10, le=3600)


class ScheduledTask(BaseModel):
    """Full task document persisted in MongoDB."""

    model_config = ConfigDict(populate_by_name=True)

    id: str = Field(..., alias="_id")
    name: str
    description: Optional[str] = None
    agent_id: str
    trigger_type: TriggerType
    trigger_config: dict
    input_payload: dict
    status: ScheduledTaskStatus = ScheduledTaskStatus.ACTIVE
    enabled: bool = True
    run_on_start: bool = False
    max_retries: int = 0
    timeout_seconds: int = 600
    owner_id: str = Field(..., description="Creator user_id")
    source_session_id: Optional[str] = None
    source_run_id: Optional[str] = None
    created_by: str = "user"
    last_run_at: Optional[datetime] = None
    last_run_status: Optional[RunStatus] = None
    last_run_id: Optional[str] = None
    total_runs: int = 0
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# ── Run record models ──────────────────────────────


class TaskRunRecord(BaseModel):
    """Single execution record persisted in MongoDB."""

    model_config = ConfigDict(populate_by_name=True)

    id: str = Field(..., alias="_id", description="run_id (UUID)")
    task_id: str
    agent_id: str
    trigger_type: str = Field("cron", description="Trigger mode: cron / interval / date / manual")
    status: RunStatus = RunStatus.PENDING
    session_id: Optional[str] = None
    trace_id: Optional[str] = None
    input_snapshot: dict = Field(default_factory=dict)
    output_result: Any = None
    error_message: Optional[str] = None
    retry_count: int = 0
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    created_at: Optional[datetime] = None


# ── API responses ──────────────────────────────────


class ScheduledTaskResponse(BaseModel):
    """API response for a scheduled task."""

    id: str
    name: str
    description: Optional[str] = None
    agent_id: str
    trigger_type: TriggerType
    trigger_config: dict
    input_payload: dict
    status: ScheduledTaskStatus
    enabled: bool
    run_on_start: bool
    max_retries: int
    timeout_seconds: int
    owner_id: str
    source_session_id: Optional[str] = None
    source_run_id: Optional[str] = None
    created_by: str = "user"
    last_run_at: Optional[datetime] = None
    last_run_status: Optional[RunStatus] = None
    last_run_id: Optional[str] = None
    total_runs: int = 0
    unread_count: int = 0
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class TaskRunResponse(BaseModel):
    """API response for a single task run."""

    id: str
    task_id: str
    agent_id: str
    trigger_type: str
    status: RunStatus
    session_id: Optional[str] = None
    trace_id: Optional[str] = None
    input_snapshot: dict
    output_result: Any = None
    error_message: Optional[str] = None
    retry_count: int = 0
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    created_at: Optional[datetime] = None


class TaskRunListResponse(BaseModel):
    """API response for paginated task run list."""

    items: list[TaskRunResponse]
    total: int


class ScheduledTaskListResponse(BaseModel):
    """API response for paginated scheduled task list."""

    items: list[ScheduledTaskResponse]
    total: int


# ── Task session responses ──────────────────────────


class TaskSessionResponse(BaseModel):
    """Lightweight session response for the scheduled-task drill-down."""

    id: str
    name: Optional[str] = None
    agent_id: str = "default"
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    is_active: bool = True
    metadata: dict[str, Any] = Field(default_factory=dict)
    unread_count: int = 0


class TaskSessionListResponse(BaseModel):
    """API response for paginated task session list."""

    items: list[TaskSessionResponse]
    total: int

File: frontend/src/types/scheduledTask.ts
// ============================================
// Scheduled Task Types
// ============================================

export type TriggerType = "interval" | "cron" | "date";
export type ScheduledTaskStatus = "active" | "paused" | "deleted";
export type ScheduledTaskCreatedBy = "user" | "agent" | "api";
export type RunStatus =
  | "pending"
  | "running"
  | "success"
  | "failed"
  | "skipped"
  | "timeout";

// Trigger configs
export interface IntervalTriggerConfig {
  seconds: number;
}

export interface CronTriggerConfig {
  year?: string | null;
  month?: string | null;
  day?: string | null;
  week?: string | null;
  day_of_week?: string | null;
  hour?: string | null;
  minute?: string | null;
  second?: string | null;
}

export interface DateTriggerConfig {
  run_date: string;
}

// Scheduled task (full response)
export interface ScheduledTask {
  id: string;
  name: string;
  description: string | null;
  agent_id: string;
  trigger_type: TriggerType;
  trigger_config: IntervalTriggerConfig | CronTriggerConfig | DateTriggerConfig;
  input_payload: Record<string, unknown>;
  status: ScheduledTaskStatus;
  enabled: boolean;
  run_on_start: boolean;
  max_retries: number;
  timeout_seconds: number;
  owner_id: string;
  source_session_id: string | null;
  source_run_id: string | null;
  created_by: ScheduledTaskCreatedBy;
  last_run_at: string | null;
  last_run_status: RunStatus | null;
  last_run_id: string | null;
  total_runs: number;
  unread_count: number;
  created_at: string | null;
  updated_at: string | null;
}

// Create request
export interface ScheduledTaskCreate {
  name: string;
  agent_id: string;
  trigger_type: TriggerType;
  trigger_config: Record<string, unknown>;
  input_payload?: Record<string, unknown>;
  description?: string | null;
  enabled?: boolean;
  run_on_start?: boolean;
  max_retries?: number;
  timeout_seconds?: number;
  source_session_id?: string | null;
  source_run_id?: string | null;
  created_by?: ScheduledTaskCreatedBy;
}

// Update request
export interface ScheduledTaskUpdate {
  name?: string;
  agent_id?: string;
  trigger_type?: TriggerType;
  trigger_config?: Record<string, unknown>;
  input_payload?: Record<string, unknown>;
  description?: string | null;
  enabled?: boolean;
  run_on_start?: boolean;
  max_retries?: number;
  timeout_seconds?: number;
}

// Task run
export interface TaskRun {
  id: string;
  task_id: string;
  agent_id: string;
  trigger_type: string;
  status: RunStatus;
  session_id: string | null;
  trace_id: string | null;
  input_snapshot: Record<string, unknown>;
  output_result: unknown;
  error_message: string | null;
  retry_count: number;
  started_at: string | null;
  finished_at: string | null;
  duration_ms: number | null;
  created_at: string | null;
}

// Paginated responses
export interface ScheduledTaskListResponse {
  items: ScheduledTask[];
  total: number;
}

export interface TaskRunListResponse {
  items: TaskRun[];
  total: number;
}

// Task session (lightweight session from scheduled task)
export interface TaskSession {
  id: string;
  name: string | null;
  agent_id: string;
  created_at: string | null;
  updated_at: string | null;
  is_active: boolean;
  metadata: Record<string, unknown>;
  unread_count: number;
}

// Paginated session list response
export interface TaskSessionListResponse {
  items: TaskSession[];
  total: number;
}
